File: elixir/easy/1704/1704.exs
defmodule Solution do
  @spec halves_are_alike(s :: String.t()) :: boolean
  def halves_are_alike(s) do
    [a, b] =
      String.split_at(s, div(String.length(s), 2))
      |> Tuple.to_list()
      |> Enum.map(&Solution.get_vowels_count/1)

    a == b
  end

  @spec get_vowels_count(string :: String.t()) :: Integer
  def get_vowels_count(string) do
    vowels_set =
      String.to_charlist("aeiouAEIOU")
      |> MapSet.new()

    string
    |> String.to_charlist()
    |> Enum.filter(fn x -> x in vowels_set end)
    |> length()
  end
end

ExUnit.start()

defmodule AssertionTest do
  use ExUnit.Case, async: true

  test "test get_vowels_count" do
    assert Solution.get_vowels_count("book") == 2
  end

  test "test_1" do
    assert Solution.halves_are_alike("book") == true
  end

  test "test_2" do
    assert Solution.halves_are_alike("textbook") == false
  end
end

File: py/easy/1704/1704.py
from functools import reduce


class Solution:
    def halvesAreAlike(self, s: str) -> bool:
        def _get_vowels_count(string: str) -> int:
            vowels = set("aeiouAEIOU")
            filtered = filter(lambda x: x in vowels, string)
            return reduce(lambda acc, x: acc + 1, filtered, 0)

        n = len(s) // 2
        return _get_vowels_count(s[:n]) == _get_vowels_count(s[n:])


solution = Solution()

assert solution.halvesAreAlike("book") is True
assert solution.halvesAreAlike("textbook") is False
